test/e2e/storage/testsuites/volumemode.go

```diff
@@ -247,10 +250,31 @@ func (t *volumeModeTestSuite) defineTests(driver TestDriver, pattern testpattern
 			framework.ExpectNoError(framework.DeletePodWithWait(f, l.cs, pod))
 		}()
 
-		// TODO: find a faster way how to check the pod can't start,
-		// perhaps when https://github.com/kubernetes/kubernetes/issues/79794 is fixed.
-		err = e2epod.WaitTimeoutForPodRunningInNamespace(l.cs, pod.Name, l.ns.Name, framework.PodStartTimeout)
-		framework.ExpectError(err, "pod with mismatched block/filesystem volumes should not start")
+		ginkgo.By("Waiting for pod to fail")
+		// Wait for an event that the pod is invalid.
+		eventSelector := fields.Set{
+			"involvedObject.kind":      "Pod",
+			"involvedObject.name":      pod.Name,
+			"involvedObject.namespace": l.ns.Name,
+			"reason":                   events.FailedMountVolume,
+		}.AsSelector().String()
+
+		var msg string
+		if pattern.VolMode == v1.PersistentVolumeBlock {
+			msg = fmt.Sprintf("has volumeMode Block, but is specified in volumeMounts")
+		} else {
+			msg = fmt.Sprintf("has volumeMode Filesystem, but is specified in volumeDevices")
+		}
+		err = e2epod.WaitTimeoutForPodEvent(l.cs, pod.Name, l.ns.Name, eventSelector, msg, framework.PodStartTimeout)
+		// Events are unreliable, don't depend on the event. It's used only to speed up the test.
+		if err != nil {
+			e2elog.Logf("Warning: did not get event about mismatched volume use")
+		}
+
+		// Check the pod is still not running
+		p, err := l.cs.CoreV1().Pods(l.ns.Name).Get(pod.Name, metav1.GetOptions{})
+		framework.ExpectError(err, "could not re-read the pod after event (or timeout)")
+		gomega.Expect(p.Status.Phase).To(gomega.Equal(v1.PodPending))
```

Comment:
It would be better to replace with `framework.ExpectEqual(p.Status.Phase, v1.PodPending` ...

Reply:
fixed